diag_manager/fms_diag_file_object.F90

```diff
@@ -582,5 +585,149 @@ subroutine add_start_time(this, start_time)
+  if (diag_file%has_file_new_file_freq()) then
+    !< If using a wildcard file name (i.e ocn%4yr%2mo%2dy%2hr), get the basename (i.e ocn)
+    pos = INDEX(diag_file_name, '%')
+    if (pos > 0) base_name = diag_file_name(1:pos-1)
+    suffix = get_time_string(diag_file_name, time_step) !TODO fname_time?
+    base_name = trim(base_name)//trim(suffix)
+  else
+    base_name = trim(diag_file_name)
+  endif
+
+  !< Add the ens number to the file name (if it exists)
+  file_name = trim(base_name)
+  call get_instance_filename(base_name, file_name)
+
+  !< Prepend the file start_time to the file name if prepend_date == .TRUE. in
+  !! the namelist
+  IF ( prepend_date ) THEN
+    call get_date(diag_file%start_time, year, month, day, hour, minute, second)
+    write (start_date, '(1I20.4, 2I2.2)') year, month, day
+
+    file_name = TRIM(adjustl(start_date))//'.'//TRIM(file_name)
+  END IF
+
+  file_name = trim(file_name)//".nc"
+
+  !< If this is a regional file add the PE and the tile_number to the filename
+  if (is_regional) then
+    !< Get the pe number that will be appended to the end of the file
+    write(mype_string,'(I0.4)') mpp_pe()
+
+    !< Add the tile number if appropriate
+    select type (domain)
+    type is (DIAGDOMAIN2D_T)
+      temp_name = file_name
+      call get_mosaic_tile_file(temp_name, file_name, .true., domain%domain2)
+    end select
+
+    file_name = trim(file_name)//"."//trim(mype_string)
+  endif
+
+  !< Open the file!
+  select type (fileobj => diag_file%fileobj)
+  type is (FmsNetcdfFile_t)
+    if (is_regional) then
+      if (.not. open_file(fileobj, file_name, "overwrite", pelist=(/mpp_pe()/))) &
+      &call mpp_error(FATAL, "Error opening the file:"//file_name)
+   else
+      allocate(pes(mpp_npes()))
+      call mpp_get_current_pelist(pes)
+
+      if (.not. open_file(fileobj, file_name, "overwrite", pelist=pes)) &
+      &call mpp_error(FATAL, "Error opening the file:"//file_name)
+   endif
+  type is (FmsNetcdfDomainFile_t)
+    select type (domain)
+    type is (diagDomain2d_t)
+      if (.not. open_file(fileobj, file_name, "overwrite", domain%Domain2)) &
+        &call mpp_error(FATAL, "Error opening the file:"//file_name)
+    end select
+  type is (FmsNetcdfUnstructuredDomainFile_t)
+    select type (domain)
+    type is (diagDomainUg_t)
+      if (.not. open_file(fileobj, file_name, "overwrite", domain%DomainUG)) &
+        &call mpp_error(FATAL, "Error opening the file:"//file_name)
+    end select
+  end select
+
+  if (diag_file%has_file_new_file_freq()) then
+    diag_file%next_open = diag_time_inc(diag_file%next_open, diag_file%get_file_new_file_freq(), &
+                                        diag_file%get_file_new_file_freq_units())
+  else
+    diag_file%next_open = diag_time_inc(diag_file%next_open, VERY_LARGE_FILE_FREQ, DIAG_DAYS)
+  endif
+
+!TODO: closing the file here for now, just to see if it works
+  call close_file(diag_file%fileobj)
```

Comment:
For the future of close_file... would it make more sense to have something like 
```fortran
if (.not.diag_file%close_file()) mpp_error("error message", fatal)
```
And have a type bound procedure that calls fms2_io close_file.